File: Rust/Masters/src/db/order.rs
// src/handlers/order_handler.rs
use crate::models::{order::{OrderRequest, OrderResponse, OrderItem}, coupon::Coupon};
use crate::db::dynamodb::get_dynamodb_client;
use crate::db::coupon::get_coupon_by_code;
use crate::db::item::get_items_by_ids;
use crate::errors::errors::OrderProcessingError;
use actix_web::{web, HttpResponse, Responder, ResponseError};
use uuid::Uuid;
use serde::{Serialize, Deserialize};
use serde_json::json;

pub async fn apply_coupon_to_order(
    order_request: web::Json<OrderRequest>,
) -> Result<OrderResponse, OrderProcessingError> {
    let client = get_dynamodb_client().await;

    // Generate a unique order ID
    let order_id = Uuid::new_v4().to_string();

    // Fetch items from the database
    let item_ids: Vec<String> = order_request.items.iter().map(|item| item.item_id.clone()).collect();
    let items = get_items_by_ids(&client, &item_ids)
    .await
    .map_err(|_| OrderProcessingError::ItemFetchError)?;

    // Calculate the total price of items in the order
    let total_price: f64 = items.iter()
        .zip(order_request.items.iter())
        .map(|(item, order_item)| item.price * order_item.quantity as f64)
        .sum();

    // Attempt to apply the coupon if provided
    let (discount_applied, final_price) = if let Some(coupon_code) = &order_request.coupon_code {
        match get_coupon_by_code(coupon_code).await {
            Ok(coupon) => {
                if coupon_applies_to_items(&coupon, &order_request.items) {
                    let discount_amount = (coupon.discount as f64) / 100.0 * total_price;
                    (true, total_price - discount_amount)
                } else {
                    return Err(OrderProcessingError::CouponNotApplicable);
                }
            }
            Err(_) => return Err(OrderProcessingError::InvalidCouponError),
        }
    } else {
        (false, total_price)
    };

    // Create the order response
    let order_response = OrderResponse {
        order_id,
        total_price,
        discount_applied,
        final_price,
    };

    // Return the order response directly
    Ok(order_response)
}


// Helper function to check if the coupon applies to the items
fn coupon_applies_to_items(coupon: &Coupon, items: &[OrderItem]) -> bool {
    if let Some(applicable_items) = &coupon.applicable_items {
        items.iter().any(|item| applicable_items.contains(&item.item_id))
    } else {
        true // If no applicable items are specified, assume the coupon is applicable to all items
    }
}

File: Python/database/order.py
import datetime
import json
from models.coupon import Coupon
from fastapi import HTTPException
from fastapi.responses import JSONResponse
from models.order import OrderRequest, OrderResponse
from .coupon import get_coupon
from .item import get_item
from typing import Optional, List
from uuid import uuid4

def calculatePrice(order: OrderRequest, coupon: Optional[Coupon], fetched_items: List[dict]) -> float:
    total_price = sum(item['price'] * order_item.quantity for item, order_item in zip(fetched_items, order.items))
    
    if not coupon:
        return total_price

    # Check if the coupon is expire
    if coupon.expires_at and datetime.datetime.fromisoformat(coupon.expires_at) < datetime.datetime.now():
        raise HTTPException(status_code=400, detail="Coupon has expired")

    discount = 0.0

    # Calculate the discount
    if coupon.applicable_items:
        for item, order_item in zip(fetched_items, order.items):
            if item['id'] in coupon.applicable_items:
                discount += item['price'] * order_item.quantity * (coupon.discount_percentage / 100)

    final_price = max(0.0, total_price - discount)  # Final price should not be negative
    return final_price

async def applyCoupon(order: OrderRequest):
    coupon = None
    if order.coupon_code:
        response = await get_coupon(order.coupon_code)
      
        if response.status_code == 200:
            response_data = json.loads(response.body.decode())
            coupon = Coupon(**response_data['Item'])
        else:
            raise HTTPException(status_code=400, detail="Coupon not found")

    fetched_items = []
    for item in order.items:
        item_response = await get_item(item.item_id)
        item_data = json.loads(item_response.body.decode())
        fetched_items.append(item_data)

    total_price = sum(item['price'] * order_item.quantity for item, order_item in zip(fetched_items, order.items))
    final_price = calculatePrice(order, coupon, fetched_items)

    order_response = OrderResponse(
        order_id = str(uuid4()),
        total_price = total_price,
        discount_applied = bool(coupon),
        final_price = final_price
    )

    return JSONResponse(content=order_response.dict() , status_code=200)

